### ISSUE TITLE
[KTS] Pre-release classes were found in dependencies. Remove them from the classpath, recompile with a release compiler or use '-Xskip-prerelease-check' to suppress errors

### ISSUE THREAD

**Issue by chachako** [gradle/gradle#21436]
The previously written **Gradle plugin** suddenly stopped working after recompiling (I'm not sure what happened during this period).

This is a confusing bug report. I have tried many methods, including cleaning the project and deleting the `.gradle` folder in the project and user home, compiling with the command line, and using the new Gradle and the old Kotlin Gradle plugin. However, none of them had any effect, which frustrated me, I had no idea where to go.

If you need more information, I will submit it as soon as possible.

## Environments
- **Gradle Wrapper Version:** `7.5` and `7.6-20220802221256+0000` (Both tried)
- **Kotlin Gradle Plugin Version:** `1.7.10` and `1.6.20` (Both tried)

## Plugin script
```kotlin
plugins { `kotlin-dsl` }

tasks.compileKotlin {
  kotlinOptions {
    freeCompilerArgs = freeCompilerArgs + "-Xskip-prerelease-check"
  }
}

dependencies {
  api(libs.square.kotlinpoet)
  compileOnly(libs.kotlin.gradle.plugin)
}
```

## Errors
```
Script compilation errors:

  Line 9:     property<String>(
              ^ Class 'com.meowool.sweekt.kotlinpoet.dsl.PropertyContainer' is compiled by a pre-release version of Kotlin and cannot be loaded by this version of the compiler

  Line 9:     property<String>(
              ^ Class 'com.meowool.sweekt.kotlinpoet.dsl.PropertysKt' is compiled by a pre-release version of Kotlin and cannot be loaded by this version of the compiler

  Pre-release classes were found in dependencies. Remove them from the classpath, recompile with a release compiler or use '-Xskip-prerelease-check' to suppress errors
```

> Wrong called function
```kotlin
inline fun <reified T> PropertyContainer.property(
  name: String,
  initializer: String,
  vararg modifiers: KModifier,
  block: Property.() -> Unit = {},
) = ...
```

## Screenshot
<img width="953" alt="image" src="https://user-images.githubusercontent.com/58068445/182697833-d522c083-59bc-4d91-a319-cdd41365df2c.png">

**Comment by ljacomet** [comment]
The problem is that you are using a library in your plugin that seems to have been compiled with a pre-release Kotlin compiler - `libs.square.kotlinpoet`. And you expose types from this library into your plugin DSL.
So while you can make your plugin compilation succeed by passing the flag, it will fail when you attempt to use the plugin and its DSL with kotlinpoet types in a project buildscript.

So this is actually a feature request for adding this flag to Gradle script compilation, as an opt-in or maybe always.

Also, it would be really helpful if you could share a full reproducer to allow us to verify the assumptions above.

**Comment by chachako** [comment]
> The problem is that you are using a library in your plugin that seems to have been compiled with a pre-release Kotlin compiler - libs.square.kotlinpoet. And you expose types from this library into your plugin DSL.
So while you can make your plugin compilation succeed by passing the flag, it will fail when you attempt to use the plugin and its DSL with kotlinpoet types in a project buildscript.

In fact, I didn't use the members of the kotlinpoet library, but compiled it myself, but obviously, it failed.

> So this is actually a feature request for adding this flag to Gradle script compilation, as an opt-in or maybe always.

And I also tried the fork Gradle source code and added the flag to the [KotlinCompiler](https://github.com/gradle/gradle/blob/v7.5.1/subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/support/KotlinCompiler.kt#L353-L356) and [KotlinBuildScript](https://github.com/gradle/gradle/blob/v7.5.1/subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/KotlinBuildScript.kt#L43-L50), but it also failed.

> Also, it would be really helpful if you could share a full reproducer to allow us to verify the assumptions above.

Because of dependency-hell, it is not easy to share a full reproducer... But I think I will try.

---

I have made various attempts in the past two days, and finally, after republishing the artifacts one by one, it finally worked! But it's frustrating that I still can't find out the cause of this mistake...

**Comment by chachako** [comment]
And I would like to add that in the process, I also found that this error would also occur if `context-receivers` was enabled for the plugin, even if I used **Gradle version 7.6** that supports **Kotlin 1.7** (but this is not the reason why I first encountered this problem).

**Comment by eskatos** [comment]
This was fixed for Gradle 8.1 as part of
* https://github.com/gradle/gradle/pull/23956

The standalone script compilation is now configured to skip the pre-release check in order to allow referencing Kotlin code compiled with more recent Kotlin language versions on a best effort basis.
